src/Cardano/Wallet/AddressDerivation.hs

```diff
@@ -0,0 +1,304 @@
+generateKeyFromSeed = unsafeGenerateKeyFromSeed
+
+-- | Derives account private key from the given root private key, using
+-- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+-- package for more details).
+--
+-- NOTE: The caller is expected to provide the corresponding passphrase.
+deriveAccountPrivateKey
+    :: Passphrase "encryption"
+    -> Key 'RootK XPrv
+    -> Index 'Hardened 'AccountK
+    -> Key 'AccountK XPrv
+deriveAccountPrivateKey (Passphrase pwd) (Key rootXPrv) (Index accIx) =
+    let
+        purposeXPrv = -- lvl1 derivation; hardened derivation of purpose'
+            deriveXPrv DerivationScheme2 pwd rootXPrv purposeIndex
+        coinTypeXPrv = -- lvl2 derivation; hardened derivation of coin_type'
+            deriveXPrv DerivationScheme2 pwd purposeXPrv coinTypeIndex
+        acctXPrv = -- lvl3 derivation; hardened derivation of account' index
+            deriveXPrv DerivationScheme2 pwd coinTypeXPrv accIx
+    in
+        Key acctXPrv
+
+-- | Derives address private key from the given account private key, using
+-- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+-- package for more details).
+--
+-- NOTE: The caller is expected to provide the corresponding passphrase. It is
+-- preferred to use 'deriveAddressPublicKey' whenever possible to avoid having
+-- to manipulate passphrases and private keys.
+deriveAddressPrivateKey
+    :: Passphrase "encryption"
+    -> Key 'AccountK XPrv
+    -> ChangeChain
+    -> Index 'Soft 'AddressK
+    -> Key 'AddressK XPrv
+deriveAddressPrivateKey (Passphrase pwd) (Key accXPrv) changeChain (Index addrIx) =
+    let
+        changeCode =
+            fromIntegral $ fromEnum changeChain
+        changeXPrv = -- lvl4 derivation; soft derivation of change chain
+            deriveXPrv DerivationScheme2 pwd accXPrv changeCode
+        addrXPrv = -- lvl5 derivation; soft derivation of address index
+            deriveXPrv DerivationScheme2 pwd changeXPrv addrIx
+    in
+        Key addrXPrv
+
+-- | Derives address public key from the given account public key, using
+-- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+-- package for more details).
+--
+-- This is the preferred way of deriving new sequential address public keys.
+deriveAddressPublicKey
+    :: Key 'AccountK XPub
+    -> ChangeChain
+    -> Index 'Soft 'AddressK
+    -> Key 'AddressK XPub
+deriveAddressPublicKey (Key accXPub) changeChain (Index addrIx) =
+    fromMaybe errWrongIndex $ do
+        let changeCode = fromIntegral $ fromEnum changeChain
+        changeXPub <- -- lvl4 derivation in bip44 is derivation of change chain
+            deriveXPub DerivationScheme2 accXPub changeCode
+        addrXPub <- -- lvl5 derivation in bip44 is derivation of address chain
+            deriveXPub DerivationScheme2 changeXPub addrIx
+        return $ Key addrXPub
+  where
+    errWrongIndex = error $
```

Comment:
So it's ok to crash with `error` when the index is greater than a billion or so?

Reply:
In practice it shouldn't happen because we _can't_ create such index. Our `Index` type is completely opaque and the only way to create indexes is through the `Bounded` and `Enum` instance (Enum which also throws at runtime if `succ` is called on `maxBound`, as specified in the base built-in). Again, the caller would have to check that prior to calling those primitives whether it makes sense.

We know it can't happen, but it's better to have a rather clear invariant here than doing a partial pattern match like `Just res <- ... `, which would not be very helpful if we ever violate this invariant.